sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,372 @@
+    for (String experiment : experiments == null ? Collections.<String>emptyList() : experiments) {
+      if (experiment.startsWith(DATA_BUFFER_SIZE_LIMIT)) {
+        return Integer.parseInt(experiment.substring(DATA_BUFFER_SIZE_LIMIT.length()));
+      }
+    }
+    return DEFAULT_BUFFER_LIMIT_BYTES;
+  }
+
+  private static long getTimeLimit(PipelineOptions options) {
+    List<String> experiments = options.as(ExperimentalOptions.class).getExperiments();
+    for (String experiment : experiments == null ? Collections.<String>emptyList() : experiments) {
+      if (experiment.startsWith(DATA_BUFFER_TIME_LIMIT_MS)) {
+        return Long.parseLong(experiment.substring(DATA_BUFFER_TIME_LIMIT_MS.length()));
+      }
+    }
+    return DEFAULT_BUFFER_LIMIT_TIME_MS;
+  }
+
+  private class Receiver<T> implements FnDataReceiver<T> {
+    private final ByteString.Output output;
+    private final Coder<T> coder;
+    private long byteCount;
+    private long elementCount;
+
+    public Receiver(Coder<T> coder) {
+      this.output = ByteString.newOutput();
+      this.coder = coder;
+      this.byteCount = 0L;
+    }
+
+    @Override
+    public void accept(T input) throws Exception {
+      int size = output.size();
+      coder.encode(input, output);
+      if (output.size() - size == 0) {
+        output.write(0);
+      }
+      final long delta = (long) output.size() - size;
+      totalByteCounter.add(delta);
+      byteCount += delta;
+      elementCount += 1;
+      if (totalByteCounter.longValue() > sizeLimit) {
+        flush();
+      }
+    }
+
+    public ByteString.Output getOutput() {
+      return output;
+    }
+
+    public void reset() {
+      this.byteCount = 0L;
+      this.elementCount = 0L;
+      this.output.reset();
+    }
+
+    @Override
+    public String toString() {
+      return String.format("Byte size: %s, Element count: %s", byteCount, elementCount);
+    }
+  }
+
+  private static class TimerEndpoint {
+
+    private final String pTransformId;
+    private final String timerFamilyId;
+
+    public TimerEndpoint(String pTransformId, String timerFamilyId) {
+      this.pTransformId = pTransformId;
+      this.timerFamilyId = timerFamilyId;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (!(o instanceof TimerEndpoint)) {
+        return false;
+      }
+      TimerEndpoint that = (TimerEndpoint) o;
+      return pTransformId.equals(that.pTransformId) && timerFamilyId.equals(that.timerFamilyId);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(pTransformId, timerFamilyId);
+    }
+
+    @Override
+    public String toString() {
+      return "TimerEndpoint{"
+          + "pTransformId='"
+          + pTransformId
+          + '\''
+          + ", timerFamilyId='"
+          + timerFamilyId
+          + '\''
+          + '}';
```

Comment:
So it looks better in the debug log.
```suggestion
      return "pTransformId: " + pTransformId + " timerFamilyId: " + timerFamilyId;
```